// react-dynamic-form/src/widgets/DynamicForm.jsx
import React from 'react';
import FormItem from './FormItem.jsx';
const splitArray = function (arr, split) {
    let currPos = 0;
    let toRet = [];
    while (currPos < arr.length) {
        let sub = [];
        for (let i = 0; i < split; i++) {
            arr[currPos] && sub.push(arr[currPos]);
            currPos++;
        }
        sub.length > 0 && toRet.push(sub);
    }
    return toRet;
};
export default class DynamicForm extends React.Component {
    constructor(props) {
        super(props);
        this.numCols = 3;//Should be divisible by 12 for proper rendering
        this.config = this.props.config;
        this.primaryKey = this.config.primaryKey;
        this.fields = this.config.fields;
        this.items = {};
        this.isEditing = false;
    }

    componentWillReceiveProps(props) {

    }

    componentDidMount() {
    }

    componentWillUnmount() {
    }

    render() {
        return (
            <div className="panel panel-default " style={{
                margin: 0,
                height: '100%',
                width: '100%'
            }}>
                <div className="panel-body">
                    <form role="form">
                        {
                            splitArray(this.fields, this.numCols).map((grp)=> {
                                const width = 12 / this.numCols;
                                return (
                                    <div key={grp[0].name} className="row">
                                        {grp.map((item)=> {
                                            return (
                                                <div key={item.name}
                                                     className={"col-sm-" + width + " col-md-" + width + " col-xs-12"}>
                                                    <FormItem ref={(itm)=> {
                                                        this.items[item.name] = itm
                                                    }} config={ item }/>
                                                </div>
                                            )
                                        })}
                                    </div>
                                )
                            })
                        }
                    </form>
                </div>
            </div>

        );
    }
}
