// react-dynamic-form/src/widgets/FormItem.jsx
import React from "react";
import Datetime from "react-datetime";
import moment from "moment";
import "react-datetime/css/react-datetime.css";

export default class FormItem extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            value: this.props.config.value,
            hasError: false
        };
        this.isRequired = this.props.config.isRequired
    }

    hasErrors(value) {
        if (this.props.config.pattern) {
            const re = new RegExp(this.props.config.pattern);
            return value.match(re) == null;
        } else {
            return false;
        }
    }

    onValueChanged(event) {
        this.setState({
            value: event.target.value,
            hasError: this.hasErrors(event.target.value)
        });
    }

    onDateValueChanged(date) {
        console.log(date);
        const dateValue = moment(date).format("YYYY-MM-DD");
        console.log("Value to set", dateValue);
        this.setState({
            value: dateValue,
            hasError: this.hasErrors(dateValue)
        });
    }

    onDateTimeValueChanged(date) {
        console.log(date);
        const dateTime = moment(date).format("YYYY-MM-DD hh:mm:ss");
        console.log("Value to set>>", dateTime);
        this.setState({
            value: dateTime,
            hasError: this.hasErrors(dateTime)
        });
    }

    render() {
        const config = this.props.config;
        switch (config.type) {
            case "number":
                return (
                    <div className={"form-group" + (this.state.hasError ? " has-error" : "")}>
                        <label htmlFor={config.name}>{config.title}</label>
                        <input type="number" className="form-control" name={config.name} value={this.state.value}
                               onChange={this.onValueChanged.bind(this)}
                               placeholder={config.placeholder}
                        />
                    </div>
                );
            case "password":
                return (
                    <div className={"form-group" + (this.state.hasError ? " has-error" : "")}>
                        <label htmlFor={config.name}>{config.title}</label>
                        <input type="password" className="form-control" name={config.name} value={this.state.value}
                               onChange={this.onValueChanged.bind(this)}
                               placeholder={config.placeholder}
                        />
                    </div>
                );
            case "checkbox":
                return (
                    <div className={"checkbox" + (this.state.hasError ? " has-error" : "")}>
                        <label><input type="checkbox" name={config.name}
                                      onChange={this.onValueChanged.bind(this)}
                                      value={this.state.value}

                        />{config.title}</label>
                    </div>
                );
            case "textarea":
                return (
                    <div className={"form-group" + (this.state.hasError ? " has-error" : "")}>
                        <label htmlFor={config.name}>{config.title}</label>
                        <textarea className="form-control" name={config.name} value={this.state.value}
                                  onChange={this.onValueChanged.bind(this)}
                                  placeholder={config.placeholder}
                        />
                    </div>
                );
            case "datetime":
                //Simple Hack to Maintain SQL Format for Dates
                this.state.value = this.state.value && moment(this.state.value).format("YYYY-MM-DD hh:mm:ss");
                return (
                    <div className={"form-group" + (this.state.hasError ? " has-error" : "")}>
                        <label htmlFor={config.name}>{config.title}</label>
                        <Datetime name={config.name} value={this.state.value}
                                  onChange={this.onDateTimeValueChanged.bind(this)}
                                  dateFormat="YYYY-MM-DD"
                                  timeFormat="hh:mm:ss"
                                  closeOnSelect={true}
                        />
                        {/* More props, refer to: https://github.com/YouCanBookMe/react-datetime */}
                    </div>
                );
            case "date":
                //Simple Hack to Maintain SQL Format for Dates...
                this.state.value = this.state.value && moment(this.state.value).format("YYYY-MM-DD");
                return (
                    <div className={"form-group" + (this.state.hasError ? " has-error" : "")}>
                        <label htmlFor={config.name}>{config.title}</label>
                        <Datetime name={config.name} value={this.state.value}
                                  onChange={this.onDateValueChanged.bind(this)}
                                  dateFormat="YYYY-MM-DD"
                                  timeFormat={false}
                                  closeOnSelect={true}
                        />
                        {/* More props, refer to https://github.com/YouCanBookMe/react-datetime */}
                    </div>
                );
            case "select":
                return (
                    <div className={"form-group" + (this.state.hasError ? " has-error" : "")}>
                        <label htmlFor={config.name}>{config.title}</label>
                        <select className="form-control selectpicker" name={config.name} value={this.state.value}
                                onChange={this.onValueChanged.bind(this)}
                        >
                            <option value="">Select...</option>
                            {config.data.map((opt) => {
                                return <option key={opt.name} value={opt.name}>{opt.title}</option>
                            })}
                        </select>
                    </div>
                );
            default :
                return (
                    <div className={"form-group" + (this.state.hasError ? " has-error" : "")}>
                        <label htmlFor={config.name}>{config.title}</label>
                        <input type="text" className="form-control" name={config.name} value={this.state.value}
                               onChange={this.onValueChanged.bind(this)}
                               placeholder={config.placeholder}
                        />
                    </div>
                );
        }

    }
}
